packages/@aws-cdk/core/lib/cfn-json.ts

```diff
@@ -0,0 +1,82 @@
+import { Construct } from './construct-compat';
+import { CustomResource } from './custom-resource';
+import { CustomResourceProvider, CustomResourceProviderRuntime } from './custom-resource-provider';
+import { Reference } from './reference';
+import { IResolvable, IResolveContext } from './resolvable';
+import { Stack } from './stack';
+import { captureStackTrace } from './stack-trace';
+
+export interface CfnJsonProps {
+  /**
+   * The value to resolve. Can be any JavaScript object, including tokens and
+   * references in keys or values.
+   */
+  readonly value: any;
+}
+
+/**
+ * Captures a synthesis-time JSON object a CloudFormation reference which
+ * resolves during deployment to the resolved values of the JSON object.
+ *
+ * The main use case for this is to overcome a limitation in CloudFormation that
+ * does not allow using intrinsic functions as dictionary keys (because
+ * dictionary keys in JSON must be strings). Specifically this is common in IAM
+ * conditions such as `StringEquals: { lhs: "rhs" }` where you want "lhs" to be
+ * a reference.
+ *
+ * This object is resolvable, so it can be used as a value.
+ *
+ * This construct is backed by a custom resource.
+ */
+export class CfnJson extends Construct implements IResolvable {
```

Comment:
Okeedokee! 

If we're going to bundle custom resources with the core library to do arbitrary data manipulation, please make this a little more generic. I like it, and I think we will be able to make good use of it for other deploy-time calculations/manipulations as well.

First thing that comes to mind: slicing arrays! I've wanted to do this many times. For example,  we cannot properly get the `roleName` from the ARN `arn:...:role/Upstairs/TheBoss/Administrator` because we can't slice-and-join, only pick-and-join.

But I'd rather have 1 Lambda Function that is set up to do multiple of these manipulations, than an individual lambda function for each, otherwise it's going to be wasteful on assets and resource count in the template.

Reply:
Okay, here is what I'll do: I'll extract the provider out into a separate file and will give it a more generic name, so we can piggy back on it later if we want to add additional operations in the future. Good?

Reply:
Excellent!